### PR TITLE
Avoid booking success redirect in case of cancellation and visit from Bookings list Page

### LINKED ISSUE #5898: [CAL-549] Redirect after event deletion
###Steps to Reproduce

1. Create an event type with the following setting: Redirect on booking (Redirect to a custom URL after a successful booking)
2. Go to the canceled event page 
3. Or click the  black "Cancel Event" button

**Actual Result:** After 8 seconds, I'm redirected to the page where we thank the customer for a booking 
** Expected result** I should only be redirected after a successful booking and not if the event is canceled
11T3HzO6SS

<sub>[CAL-549](https://linear.app/calcom/issue/CAL-549/redirect-after-event-deletion)</sub>

### PR DESCRIPTION
**Pull request by hariombalhara**
## What does this PR do?

- Fixes #5898
- Also, fixes the issue where on refresh SSR of the success page was failing when Redirection toast was there.

[Loom](https://www.loom.com/share/862fc65dce52453db3e187939c1737e8)

**Environment**: Production

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## How should this be tested?

- [x] On successful booking, success redirect should work. If user clicks on Cancel in the meantime, the toast would go away and redirect would stop immediately
- [x] Visit the same success page from Bookings listing(/bookings) and redirect shouldn't occur in that case.

### CODE REVIEW

File: apps/web/pages/booking/[uid].tsx [hunk 1]
```diff
@@ -57,26 +57,40 @@ function redirectToExternalUrl(url: string) {
...
+  useEffect(() => {
+    if (!isToastVisible && timerRef.current) {
+      window.clearInterval(timerRef.current);
+    }
+  }, [isToastVisible]);
 
   useEffect(() => {
+    const parsedExternalUrl = new URL(url);
+
+    const parsedSuccessUrl = new URL(document.URL);
```
Comment (hariombalhara) [inline apps/web/pages/booking/[uid].tsx:79]: Moved parsedSuccessUrl logic to useEffect to fix SSR bug.

File: apps/web/pages/booking/[uid].tsx [hunk 2]
```diff
@@ -57,26 +57,40 @@ function redirectToExternalUrl(url: string) {
...
-    parsedSuccessUrl.searchParams.toString() +
-    parsedExternalUrl.hash;
 
   const { t } = useLocale();
   const timerRef = useRef<number | null>(null);
+  const router = useRouter();
+  const { cancel: isCancellationMode } = querySchema.parse(router.query);
+
+  if (isCancellationMode && timerRef.current) {
+    setIsToastVisible(false);
```
Comment (hariombalhara) [inline apps/web/pages/booking/[uid].tsx:67]: Redirect toast shouldn't be visible if cancellation is being requested.

File: apps/web/pages/booking/[uid].tsx [hunk 3]
```diff
@@ -57,26 +57,40 @@ function redirectToExternalUrl(url: string) {
...
+  }
+
+  useEffect(() => {
+    if (!isToastVisible && timerRef.current) {
```
Comment (hariombalhara) [inline apps/web/pages/booking/[uid].tsx:71]: If toast is not visible, redirect shouldn't happen.

File: apps/web/pages/booking/[uid].tsx [hunk 4]
```diff
@@ -124,7 +138,6 @@ function RedirectionToast({ url }: { url: string }) {
                   type="button"
                   onClick={() => {
                     setIsToastVisible(false);
-                    window.clearInterval(timerRef.current as number);
```
Comment (hariombalhara) [inline apps/web/pages/booking/[uid].tsx:127]: This is now tackled by isToastVisible check  above

File: apps/web/pages/booking/[uid].tsx [hunk 5]
```diff
@@ -57,26 +57,41 @@ function redirectToExternalUrl(url: string) {
...
+      parsedSuccessUrl.searchParams.set(name, value);
+    }
+
+    const urlWithSuccessParams =
+      parsedExternalUrl.origin +
+      parsedExternalUrl.pathname +
+      "?" +
+      parsedSuccessUrl.searchParams.toString() +
+      parsedExternalUrl.hash;
+    urlWithSuccessParamsRef.current = urlWithSuccessParams;
```
Comment (hariombalhara) [inline apps/web/pages/booking/[uid].tsx:93]: Store it in Ref so that the variable can be accessed outside useEffect.

**Review by hariombalhara — COMMENTED** [review]
Self review done

File: apps/web/pages/booking/[uid].tsx [hunk 6]
```diff
@@ -349,7 +365,9 @@ export default function Success(props: SuccessProps) {
       <CustomBranding lightVal={props.profile.brandColor} darkVal={props.profile.darkBrandColor} />
       <main className={classNames(shouldAlignCentrally ? "mx-auto" : "", isEmbed ? "" : "max-w-3xl")}>
         <div className={classNames("overflow-y-auto", isEmbed ? "" : "z-50 ")}>
-          {eventType.successRedirectUrl ? <RedirectionToast url={eventType.successRedirectUrl} /> : null}{" "}
+          {isSuccessBookingPage && !isCancellationMode && eventType.successRedirectUrl ? (
```
Comment (hariombalhara) [inline apps/web/pages/booking/[uid].tsx:368]: Shouldn't be cancellation case and should be the visit right after booking.

**Review by CarinaWolli — APPROVED** [review]
Looks good 👍

**Pull request merged**